### PR TITLE
[Feature] Ring implementation

### PR DESCRIPTION
**Pull request by duongcongtoai**
Ring uses consistent hashing with the first key of the command to determine the request location. (Multi-key command will have issue using Ring).
#16 #14

### CODE REVIEW

File: go-client/ring.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+	r.mu.RUnlock()
+	if !ok {
+		return fmt.Errorf("address after consistent hash %s does not exist", addr)
+	}
+	return correctShard.client.commandAdaptor(cmd)
+}
+
+func (r *Ring) getShards() []*shardClient {
+	var ret []*shardClient
+	r.mu.RLock()
```
Comment (ldhnam) [inline go-client/ring.go:149]: Should move `r.mu.RUnlock()` below `RLock()`, and should use `defer r.mu.RUnlock()`

File: go-client/ring.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	return r, nil
+}
+
+func (r *Ring) Close() error {
+	return nil
+	//TODO
+}
+
+func (r *Ring) cmdWithConsistentHash(cmd Cmd) error {
+	r.mu.RLock()
```
Comment (ldhnam) [inline go-client/ring.go:132]: Should move `r.mu.RUnlock()` below `RLock()`, and should you `defer r.mu.RUnlock()`

File: go-client/ring.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+			}
+		}
+		if shardsChanged {
+			r.updateAliveShards()
+		}
+	}
+}
+
+func (r *Ring) updateAliveShards() {
+	r.mu.RLock()
```
Comment (ldhnam) [inline go-client/ring.go:178]: Should move `r.mu.RUnlock()` below `RLock()`, and should you `defer r.mu.RUnlock()`

**Comment by duynguyen-ori75** [comment]
Consistent hashing is meant to be implemented in the server, not in the client

The reasons behind it are:
- Better replication management
- Finger table management (routing information stored in each node)
- Node joins/leaves/crashes

Please consider re-design based on the original consistent hashing if it is possible!

File: go-client/ring.go (hunk 2 above)
Comment (duongcongtoai) [reply go-client/ring.go:132]: This is used to release the lock early, using defer means the lock is released only after the response from keva is returned.

**Comment by tuhuynh27** [comment]
@duynguyen-ori75 Yes we acknowledge that.

However, in the first stage, we want to try to clone "Redis-Ring" feature which is a client sharding only (of course this will not be able to handle data loss due to node join/leave/crash). This is for the "simple thing comes first".

The next stage should be shard and replica on the server-side.

More context on this:

Ref: [https://engineering.linecorp.com/en/blog/line-storage-storing-billions-of-rows-in-sharded-redis-and-hbase/](https://engineering.linecorp.com/en/blog/line-storage-storing-billions-of-rows-in-sharded-redis-and-hbase/)
> The entire LINE storage system started with just a single Redis cluster constructed from 3 nodes sharded on the client-side.

In 2011, LINE storage team adopt a similar approach "client shard", they then migrate to server-sharding and replication.

**Review by tuhuynh27 — APPROVED** [review]
LGTM

File: go-client/ring.go (hunk 2 above)
Comment (ldhnam) [reply go-client/ring.go:132]: Ah oh, got it

**Comment by ldhnam** [comment]
LGTM

**Review by ldhnam — APPROVED** [review]

**Review by axblueblader — APPROVED** [review]

**Pull request merged**